engine/collection/ingest/engine.go

```diff
@@ -87,75 +126,122 @@ func New(
 	return e, nil
 }
 
-// Ready returns a ready channel that is closed once the engine has fully
-// started.
+// Start starts the engine by starting the message processing loop.
+func (e *Engine) Start(ctx irrecoverable.SignalerContext) {
+	e.lm.OnStart(func() {
+		e.wg.Add(1)
+		go e.loop(ctx)
+	})
+}
+
+// Ready returns a ready channel that is closed once the engine has fully started.
 func (e *Engine) Ready() <-chan struct{} {
-	return e.unit.Ready()
+	return e.lm.Started()
 }
 
 // Done returns a done channel that is closed once the engine has fully stopped.
 func (e *Engine) Done() <-chan struct{} {
-	return e.unit.Done()
+	e.lm.OnStop()
+	return e.lm.Stopped()
+}
+
+// Process processes the given event from the node with the given origin ID in
+// a blocking manner. It returns the potential processing error when done.
+func (e *Engine) Process(channel network.Channel, originID flow.Identifier, event interface{}) error {
+	err := e.messageHandler.Process(originID, event)
+	if err != nil {
+		if engine.IsIncompatibleInputTypeError(err) {
+			e.log.Warn().Msgf("%v delivered unsupported message %T through %v", originID, event, channel)
+			return nil
+		}
+		return fmt.Errorf("unexpected error while processing engine message: %w", err)
+	}
+	return nil
+}
+
+// ProcessLocal processes an event originating on the local node. For local messages,
+// we skip the message queue and immediately process the transaction, blocking
+// the caller until we are finished. This way, we can provide context to the
+// user about whether their transaction was accepted.
+func (e *Engine) ProcessLocal(event interface{}) error {
+	tx, ok := event.(*flow.TransactionBody)
+	if !ok {
+		return fmt.Errorf("local submission of non-transaction event type (%T) to ingest engine: %w", event, engine.IncompatibleInputTypeError)
+	}
+	e.engMetrics.MessageReceived(metrics.EngineCollectionIngest, metrics.MessageTransaction)
+	return e.onTransaction(e.me.NodeID(), tx)
 }
 
 // SubmitLocal submits an event originating on the local node.
 func (e *Engine) SubmitLocal(event interface{}) {
-	e.unit.Launch(func() {
-		err := e.process(e.me.NodeID(), event)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
+	err := e.ProcessLocal(event)
+	if err != nil {
+		e.log.Fatal().Err(err).Msg("internal error processing event")
+	}
 }
 
 // Submit submits the given event from the node with the given origin ID
 // for processing in a non-blocking manner. It returns instantly and logs
 // a potential processing error internally when done.
 func (e *Engine) Submit(channel network.Channel, originID flow.Identifier, event interface{}) {
-	e.unit.Launch(func() {
-		err := e.process(originID, event)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
+	err := e.Process(channel, originID, event)
+	if err != nil {
+		e.log.Fatal().Err(err).Msg("internal error processing event")
```

Comment:
Referring back to the comments above, this is an example where ideally you would throw an irrecoverable error using `ctx.Throw`, rather than logging fatal.